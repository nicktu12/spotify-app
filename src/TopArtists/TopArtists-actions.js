export const saveAuthCodeAction = (code) => {
  return {
    type: 'AUTH_CODE',
    code
  };
};
