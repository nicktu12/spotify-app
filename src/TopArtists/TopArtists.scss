@import '../Styles/Variables.scss';

.home-div {
	height: calc(100vh - 90px);
}

.home-div h2 {
  background-color: $pop-accent-o;
  color: $white;
  text-align: center;
  padding-top: 50px;
	height: 91px;
	box-sizing: border-box;
	font-family: $header-font;
	font-size: 32px;
}

.home-div ol {		
  background-color: $dark-bg;
  text-align: left;
	color: $light-3;
  padding-left: calc(30vw + 35px);
	overflow: scroll;
	height: calc(99vh - 174px);
}

h2 img {
	height: 20px;
	width: 20px;
}

.home-div li {
  height: 42px;
	margin: 4px 0px;
	font-family: $header-font;
	border-radius: 1px;
  box-sizing: border-box;
}

li button {
  border: none;
  background-color: rgba(0,0,0,0);
  width: 100%;
  text-align: left;
  height: 42px;
  outline-color: $pop-accent;
  font-family: $body-font;
  color: $light-3;
  font-size: 16px;
}

li button:hover {
  cursor: pointer;
}

.artists .plus-icon, .artists .minus-icon {
	display: none;
	color: $light-3;
}

.home-div li:hover {
	background-color: $dark;
  border-bottom: 2px solid $pop-accent;
}

.unselected:hover .plus-icon {
	display: inline;
}

.selected:hover .minus-icon {
	display: inline;
}

.artists span {
	display: block;
	padding-left: 10px;
}

.unselected {
	img {
		display: none;
	}
	p {
		display: none;
	}
}

.selected {
	background-color: $dark;
  border-bottom: 2px solid $pop-accent;
}

section {
  position: absolute;
  top: 181px;
  height: calc(99vh - 191px);
  width: 30vw;
  overflow: scroll;
}

section {
  h4 {
    font-family: $header-font;
    font-size: 24px;
    color: $white;
    padding: 15px;
  }
	img {
		width: 24vw;
    border: 2px solid $pop-accent;
    border-radius: $border-rad / 4;
	}
  p {
    color: $white;
    text-align: left;
    font-family: $body-font;
    line-height: 25px;
    width: 22vw;
    margin: auto;
  }
	div {
    width: 22vw;
    margin: auto;
		vertical-align: top;
		flex-direction: column;
		justify-content: space-around;
	}
	div p {
		font-family: $body-font;
    line-height: 25px;
	}
	div p span {
		font-family: $body-font;
		display: initial;
		font-size: 16px;
		color: $white;
	}
	.alt-text {
		color: $light-3;
		width: calc(100vw - 700px);
		line-height: normal;
    padding-left: 5px;
	}
}

