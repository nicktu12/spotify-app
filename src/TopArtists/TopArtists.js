import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { authCodeCleaner } from '../Utilities/helpers';
import { saveAuthCodeAction } from './TopArtists-actions'; 
import { Meter } from '../Meter/Meter';

export class TopArtists extends React.Component{

  constructor(){
    super();

    this.state = {
      selected: null,
    };
  }

  componentDidMount(){
    const url = window.location.href;
    if (url.includes('code')) {
      this.props.authCodeToSagas(authCodeCleaner(url));
    } else if (url.includes('error')){
      alert('error with url');
    } else {
      this.redirectToLogin();
    }
  }

  renderTopArtists = (array) => (
    array.map((artist, index) => (
      <li 
        key={'top artists ' + index} 
        className={ 
          this.state.selected === index ? 
            'selected artists' : 
            'unselected artists' }
      >
        <button onClick={()=>this.addSelected(index)}>
          <span>{artist.name} 
            <span className='plus-icon'>+</span>
            <span className='minus-icon'>-</span>
          </span>
        </button>
      </li>
    ))      
  ) 

  addSelected = (index) => (
    this.state.selected === null ? 
      this.setState({ selected: index}) : 
      this.updateSelected(index)
  )

  updateSelected = (index) => {
    this.state.selected === index ?
      this.setState({ selected: null }) :
      this.setState({ selected: index });
  }

  redirectToLogin = () => {
    if (!this.props.token.length) {
      this.props.history.push('/login');
    }     
  } 

  showLoading = () => ( 
    !this.props.topArtists.length && 
    <img src={require('../Assets/bars.svg')} alt="loading icon"  />
  )

  renderRecentlyPlayed = (songArray) => (
    songArray.map(song => (
      <li>
        { song.title } - { song.artists }
      </li> 
    ))
  )

  renderInfoCard = (info) => (
    info === undefined ?
      <section>
        <h4>{this.props.userInfo.name}</h4>
        {
          this.props.userInfo.image &&
          <img src={this.props.userInfo.image} alt='user' />
        }
        <p className='user-info'>
          {
            this.props.userInfo.followers &&
            <span>
              <h5>Recently Played</h5>
              <ol className='recently-played-list'>
               { this.renderRecentlyPlayed(this.props.recentlyPlayed) }
              </ol>
            </span>
          }
        </p>
      </section>
      :
      <section>
        <h4>{info.name}</h4>
        <img src={info.photo.url} alt={info.name + ' photo'} />
        <div>
          <p>
            <span>Followers:</span> <span className='alt-text'>
              {info.followers}
            </span>
          </p>
          <p>
            <span>Popularity:</span> <span className='alt-text'>
              <Meter percent={info.popularity / 100} rounded={true} />
            </span>
          </p>
          <p>
            <span>Genres:</span> <span className='genres alt-text'>
              {info.genres}
            </span>
          </p>
        </div>
      </section>
  )

  render(){
    return (
      <div className='home-div'>
        { this.renderInfoCard(this.props.topArtists[this.state.selected]) }
        <h2>{this.showLoading()} Top Artists {this.showLoading()}</h2>
        <ol>
          { this.props.topArtists && 
            this.renderTopArtists(this.props.topArtists)
          }
        </ol>
      </div>
    );
  }

}

const mapStateToProps = store => ({
  topArtists: store.topArtists,
  userInfo: store.userInfo,
  token: store.accessToken,
  recentlyPlayed: store.recentlyPlayed,
});

const mapDispatchToProps = dispatch => {
  return {
    authCodeToSagas: (token) => dispatch(saveAuthCodeAction(token))
  };
};

TopArtists.propTypes = {
  authCodeToSagas: PropTypes.func,
  token: PropTypes.string,
  history: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  topArtists: PropTypes.arrayOf(PropTypes.object),
  userInfo: PropTypes.object,
};

export default connect(mapStateToProps, mapDispatchToProps)(TopArtists);
