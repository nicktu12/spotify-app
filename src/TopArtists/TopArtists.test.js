import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import TopArtists from './TopArtists';
import { 
  mockTopArtists, 
  mockUserInfo, 
  mockAccessToken,
} from '../Utilities/mockData';


describe('TopArtists component tests', ()=>{
  it('should match snapshot render', ()=>{
    const mockFunc = jest.fn();
    const mockStore = configureStore();
    const initialState = {
      accessToken: '',
      topArtists: [],
      topSongs: [],
      topSongsShortTerm: [],
      topSongsAllTime: [],
      userInfo: {},
    };
    const store = mockStore(initialState);
    const wrapper = shallow(
      <TopArtists 
        history={[]}
        store={store}
        authCodeToSagas={mockFunc}
        topArtists={mockTopArtists}
        userInfo={mockUserInfo}
        token={mockAccessToken}
      />
    ); 

    expect(wrapper).toMatchSnapshot();
  });
});
