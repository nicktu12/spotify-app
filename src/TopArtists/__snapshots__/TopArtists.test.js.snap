// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopArtists component tests should match snapshot render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(TopArtists)
    authCodeToSagas={[Function]}
    history={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    token="BxxxxxxxU8xxxxxxxvK5xxxxxxxxxsBrtbsaM_37uIB1yaFCPApS0Iczif6KIKq8EXAwXtoQyG8FduappKkTi0JYNO-a4E7ioKBNlpMmCI9zf4KiMd9_ZN3YpFXv0TRUhCuxq_WXHLoRn-M-cSPJ5AYZdRFrB7_dwkwtLsOr"
    topArtists={
      Array [
        Object {
          "name": "Snoop Dogg",
        },
        Object {
          "name": "The Grateful Dead",
        },
        Object {
          "name": "Solange",
        },
      ]
    }
    userInfo={
      Object {
        "email": "edgarthecat@cats.com",
        "image": "https://scontent.xx.fbcdn.net/v/t1.0-1/aaaaaaaaaaaaaa558a1aaaaa0aaaaaa5063aaaa4aaaa12218136501_n.jpg?oh=53ef504a0ba139726593f4497886fa16&oe=5A721441",
        "name": "Edgar the Cat",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "authCodeToSagas": [Function],
      "history": Array [],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
      "token": "",
      "topArtists": Array [],
      "userInfo": Object {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "authCodeToSagas": [Function],
        "history": Array [],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "token": "",
        "topArtists": Array [],
        "userInfo": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
